[SandboxIR] Add InsertValueInst

slackito opened the pull request.

slackito committed:
[SandboxIR] Add InsertValueInst

slackito committed:
fix clang-format issues

tschuett commented on llvm/include/llvm/SandboxIR/SandboxIR.h:
```diff
@@ -1451,6 +1456,63 @@ class ShuffleVectorInst final
+                       BBIterator WhereIt, BasicBlock *WhereBB, Context &Ctx,
+                       const Twine &Name = "");
+
+  using idx_iterator = llvm::InsertValueInst::idx_iterator;
```
No
```
static bool classof(const Value *From) {
    return From->getSubclassID() == ClassID::InsertValueInst;
}
```

slackito replied:
@tschuett good catch, thanks!

@vporpo It does pass, and I just double-checked I'm running them with asserts enabled (debug build).

I think it's falling back to `Instruction::classof` and accidentally passing.

vporpo replied:
I wonder if there is a good way to check that classof has been implemented. Perhaps a test that uses the .def file and  does something with `<subclass>::classof()` so that we get a compile-time error if we forget to implement it.

vporpo replied:
I wrote a small test that checks `EXPECT_NE(&<subclass>::classof, &Instruction::classof);` and one of the insturctions is missing a classof!

slackito replied:
Great! Can you commit that test (and the missing classof) separately from this PR?

vporpo replied:
Yes I will upload a PR in a few minutes.

vporpo commented on llvm/unittests/SandboxIR/SandboxIRTest.cpp:
```diff
@@ -1228,6 +1228,98 @@ define void @foo(<2 x i8> %v1, <2 x i8> %v2) {
+  auto *Zero = sandboxir::ConstantInt::get(Type::getInt32Ty(C), 0, Ctx);
+  auto *ShouldBeConstant = sandboxir::InsertValueInst::create(
+      Ins1->getOperand(0), Zero, ArrayRef<unsigned>({0}), BB->end(), BB, Ctx);
+  EXPECT_TRUE(isa<sandboxir::Constant>(ShouldBeConstant));
```
nit: Constants are unique, so we could also do something `EXPECT_EQ(ShouldBeConstant, ExpectedConstant)`

slackito replied:
Done.

vporpo commented on llvm/unittests/SandboxIR/SandboxIRTest.cpp:
```diff
@@ -1228,6 +1228,98 @@ define void @foo(<2 x i8> %v1, <2 x i8> %v2) {
+  EXPECT_TRUE(isa<sandboxir::Constant>(ShouldBeConstant));
+
+  // idx_begin / idx_end
+  auto IdxIt = Ins0->idx_begin();
```
nit: Could you check the indices with an InsertValueInst with at least 2 indices?

slackito replied:
Done.

slackito commented on llvm/include/llvm/SandboxIR/SandboxIR.h:
```diff
@@ -1451,6 +1456,63 @@ class ShuffleVectorInst final
+  static Value *create(Value *Agg, Value *Val, ArrayRef<unsigned> Idxs,
+                       BBIterator WhereIt, BasicBlock *WhereBB, Context &Ctx,
+                       const Twine &Name = "");
+
```
Are you sure? I only see `getIndexedType` methods in `GetElementPtrInst` and `ExtractValueInst`, not in `InsertValueInst`.

vporpo replied:
Oops I was looking at the `ExtractValueInst` :)

slackito committed:
Address review comments.

- Added `classof` static method.
- Added test case for `insertvalue` with more than one index.
- Check constant equality in the "folded constant" case

vporpo reviewed (approved):
Looks good, thanks!
